src/main/java/com/minecolonies/coremod/colony/Structures.java

```diff
@@ -60,200 +109,845 @@ public static void init()
+     * @return the structure name as 'schematics/wooden/Builder1' or an empty String if not found
+     */
+    public static StructureName getStructureNameByMD5(final String md5)
+    {
+        if (md5 != null)
+        {
+            for (final Map.Entry<String, String> md5Entry : md5Map.entrySet())
+            {
+                if (md5Entry.getValue().equals(md5))
+                {
+                    return new StructureName(md5Entry.getKey());
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns a map of all the structures.
+     *
+     * @return List of structure with their md5 hash.
+     */
+    public static Map<String, String> getMD5s()
+    {
+        return Structures.md5Map;
+    }
+
+    /**
+     * get the set of cached schematic.
+     */
+    private static Set<String> getCachedMD5s()
+    {
+        final Set<String> md5Set = new HashSet<>();
+        for (final Map.Entry<String, String> md5 : md5Map.entrySet())
+        {
+            final StructureName sn = new StructureName(md5.getKey());
+            if (sn.getSection().equals(SCHEMATICS_CACHE))
+            {
+                md5Set.add(md5.getKey());
+            }
+        }
+        return md5Set;
+    }
+
+    /**
+     * delete a cached structure.
+     * delete the file and the md5 entry
+     *
+     * @param structureName the structure to delete
+     * @return True if the structure have been deleted, False otherwise
+     */
+    private static boolean deleteCachedStructure(@NotNull final StructureName structureName)
+    {
+        if (!SCHEMATICS_CACHE.equals(structureName.getPrefix()))
+        {
+            Log.getLogger().warn("Delete failed: Invalid name " + structureName);
+            return false;
+        }
+
+        if (!hasMD5(structureName))
+        {
+            Log.getLogger().warn("Delete failed: No MD5 hash found for " + structureName);
+            return false;
+        }
+
+        final File structureFile = Structure.getSchematicsFolder().toPath().resolve(structureName.toString() + SCHEMATIC_EXTENSION).toFile();
+        if (structureFile.delete())
+        {
+            md5Map.remove(structureName.toString());
+            return true;
+        }
+        else
+        {
+            Log.getLogger().warn("Failed to delete structure " + structureName);
+        }
+        return false;
+    }
+
+    /**
+     * check that we can store the schematic.
+     * According to the total number of schematic allowed on the server
+     *
+     * @return true if we can store more schematics
+     */
+    private static boolean canStoreNewSchematic()
+    {
+        if (MineColonies.isClient())
+        {
+            return true;
+        }
+        if (!Configurations.allowPlayerSchematics)
+        {
+            return false;
+        }
+
+        final Set<String> md5Set = getCachedMD5s();
+        Log.getLogger().info("Server has " + md5Set.size() + " cached schematics");
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "Server has " 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)